### PR TITLE
testing/pypy: new aport

### PR DESCRIPTION
**Pull request by HorridBro**
https://github.com/HorridBro/AlpinePyPy
PyPy build file

### CODE REVIEW

**Comment by HorridBro** [comment]
The reason why Travis fails is because PyPy build time exceeds  easily two hours on a regular computer.

File: testing/pypy/APKBUILD [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Contributor:Mihnea Saracin <mihnea.saracin@rinftech.com>
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: Please put a space between the colon and the names.

File: testing/pypy/APKBUILD [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+# Maintainer:Mihnea Saracin <mihnea.saracin@rinftech.com>
+pkgname=pypy
+pkgver=5.7.1
+pkgrel=0
+pkgdesc="pypy2"
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: A better package description would be nice. Maybe something like this: `pkgdesc="A fast, compliant alternative implementation of the Python 2 language"`.

File: testing/pypy/APKBUILD [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+url="https://bitbucket.org/pypy"
+arch="all"
+license="MIT"
+depends="libssl1.0 libcrypto1.0 libffi libbz2 "
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: This *should* be automatically inferred by `abuild` as noted at the end of the build log, although I can't double-check since the Travis build is timing out.

File: testing/pypy/APKBUILD [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+makedepends="gcc make python linux-headers  libffi-dev pkgconf zlib-dev bzip2-dev
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: At least `gcc`, `make`, `libc-dev` and `tk` can likely be removed from `makedepends`: The former three belong to `build-base`, and the latter *should* be implied by `tk-dev`. Also, does `pypy` not build against LibreSSL? I would suggest `libressl-dev` if it does, because we try to limit the number of packages linking against OpenSSL.

File: testing/pypy/APKBUILD [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+pkgver=5.7.1
+pkgrel=0
+pkgdesc="pypy2"
+url="https://bitbucket.org/pypy"
+arch="all"
+license="MIT"
+depends="libssl1.0 libcrypto1.0 libffi libbz2 "
+makedepends="gcc make python linux-headers  libffi-dev pkgconf zlib-dev bzip2-dev 
+            sqlite-dev ncurses-dev readline-dev openssl-dev expat-dev openssl-dev gdbm-dev tk tk-dev libc-dev py-cffi paxmark"
+install=""
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: This line can be omitted if it's empty.

File: testing/pypy/APKBUILD [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+subpackages="$pkgname-dev $pkgname-doc"
+source="https://bitbucket.org/$pkgname/$pkgname/downloads/pypy2-v$pkgver-src.tar.bz2
+        alpine.patch"
+builddir="$srcdir/pypy2-v$pkgver-src/pypy/goal"
+
+
+prepare(){
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: This entire function is part of the default prepare implementation and can be omitted.

File: testing/pypy/APKBUILD [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        for i in $source; do
+                case $i in
+                *.patch) msg $i; patch -p1 -i "$srcdir"/$i || return 1
+                esac
+        done
+}
+
+build() {
+        cd "$builddir"
+        /usr/bin/python ../../rpython/bin/rpython --opt=jit --nopax targetpypystandalone.py || return 1
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: No need for a full path here, and please explicitly the Python version (`python2` instead of `/usr/bin/python`, in short).

Comment (Shizmob) [inline testing/pypy/APKBUILD]: Also applies below: `|| return 1` can be omitted as `set -e` is default in APKBUILDs now.

File: testing/pypy/APKBUILD [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+package() {
+        mkdir -p "$pkgdir"
+        cd "$builddir/../tool/release"
+        ./package.py --archive-name pypy2-v$pkgver --builddir "$srcdir/pypy2-v$pkgver-alpine" || return 1
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: This doesn't seem to be installing PyPy to `$pkgdir`? There is no reference of `$pkgdir` in this command-line at least. Does PyPy do some magic on its own?

Comment (Shizmob) [inline testing/pypy/APKBUILD]: Also, the `|| return 1` statement can be omitted.

File: testing/pypy/APKBUILD [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        done
+}
+
+build() {
+        cd "$builddir"
+        /usr/bin/python ../../rpython/bin/rpython --opt=jit --nopax targetpypystandalone.py || return 1
+}
+
+package() {
+        mkdir -p "$pkgdir"
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: This can be removed if the install step creates it automatically, which seems to be the case for a lot of build systems. If it's not the case for PyPy it's not a problem, but yeah.

File: testing/pypy/APKBUILD [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        cd "$builddir/../tool/release"
+        ./package.py --archive-name pypy2-v$pkgver --builddir "$srcdir/pypy2-v$pkgver-alpine" || return 1
+}
+
+
+
+md5sums="73720fd65aa634d60d4633a4c4e38cbb  pypy2-v5.7.1-src.tar.bz2
```
Comment (Shizmob) [inline testing/pypy/APKBUILD]: `md5sums=` and `sha256sums=` have been deprecated and can be removed.

**Comment by Shizmob** [comment]
Furthermore, does PyPy have a testsuite? Since recently we implemented a `check()` function in APKBUILDs that runs the test suite (and installs any dependencies for it defined in `checkdepends=`) -- you may need to upgrade your build system to edge to be able to test that, but since this is a PR against edge, that is probably a good idea anyway. If there was and it ran succesfully (or mostly), that would significantly increase the confidence in a working PyPy.

**Comment by Shizmob** [comment]
Last random note: while I do love the beautiful fractals it draws, they are kind of a noise in our build log this way, making it harder to find errors. Could it be disabled somehow?

File: testing/pypy/APKBUILD (hunk 4 above)
Comment (Shizmob) [inline testing/pypy/APKBUILD]: Also, `py-cffi` should probably be changed to `py2-cffi` for clarity.

**Comment by kaniini** [comment]
I prefer not accepting pypy into Alpine because it justifies keeping python 2 around.  pypy3 would, however, be fine.

While, originally, we decided to keep python 2 around because there were many consumers, these consumers are declining and python 2 only has limited upstream maintenance.  Therefore it makes no sense to add pypy.

**Comment by HorridBro** [comment]
Hi, @kaniini  , I don't know why are you saying that python2 customers are declining since many projects are still using it (for example OpenStack). I don't see a problem in adding pypy2 alongside python2.
The customer will choose which one is the best for him.

Regarding pypy3, I'm working on an APKBUILD file for it and once it's done I'll submit a pull request.

**Comment by Shizmob** [comment]
I've got no particular issues myself with adding pypy, especially as pypy3 is somewhat experimental and existing software benefits from pypy.

**Comment by HorridBro** [comment]
I updated the files after the review.Regarding the fractal , it is suppressed , but i kept all other logging for debugging purposes .

File: testing/pypy/APKBUILD (hunk 3 above)
Comment (Shizmob) [inline testing/pypy/APKBUILD]: This still holds, I think?

File: testing/pypy/APKBUILD (hunk 4 above)
Comment (Shizmob) [inline testing/pypy/APKBUILD]: Was `tk` not implied by `tk-dev`?

**Comment by Shizmob** [comment]
Another comment: does PyPy have a testsuite we can run for `check()` implementation? Also CC'ing @ncopa and @kaniini again for opinions on inclusion.

**Comment by kaniini** [comment]
My position is NAK, as I said earlier.  We want less python 2, not more python 2.

**Comment by kaniini** [comment]
If PyPy developers care about their work being included here, they should make PyPy3 stable and concentrate on supporting 3.6.  Hard NAK for any new Python 2 interpreters.

**Comment by jirutka** [comment]
I fully agree with @kaniini. Python 2 must *finally* die.

**Comment by clandmeter** [comment]
Closed due to inactivity and reviewers NAK.
